File: BeeCrowd/1034-Ice-Statues-Festival/solution.js
function solve() {
  const fs = require("fs");
  const input = fs.readFileSync("/dev/stdin", "utf8");
  let pos = 0;

  // Custom integer parser for speed (faster than input.split)
  function readInt() {
    while (pos < input.length && input.charCodeAt(pos) <= 32) pos++;
    if (pos >= input.length) return null;
    let res = 0;
    while (pos < input.length && input.charCodeAt(pos) > 32) {
      res = res * 10 + (input.charCodeAt(pos) - 48);
      pos++;
    }
    return res;
  }

  const T = readInt();
  if (T === null) return;

  const dp = new Int32Array(12000); // Safe buffer size

  for (let t = 0; t < T; t++) {
    const N = readInt();
    let M = readInt();

    const coins = new Int32Array(N);
    for (let i = 0; i < N; i++) coins[i] = readInt();

    coins.sort((a, b) => b - a); // Sort Descending

    const maxCoin = coins[0];
    let count = 0;

    // Dynamic Limit Heuristic
    let safe_limit = 0;
    if (N >= 2) safe_limit = maxCoin * coins[1];

    // Greedy Reduction
    if (M > safe_limit) {
      const reduceAmount = M - safe_limit;
      const numMaxCoins = Math.floor(reduceAmount / maxCoin);
      if (numMaxCoins > 0) {
        count += numMaxCoins;
        M -= numMaxCoins * maxCoin;
      }
    }

    // DP on Remainder
    const INF = 1e9;
    dp[0] = 0;
    for (let i = 1; i <= M; i++) {
      dp[i] = INF;
      for (let j = 0; j < N; j++) {
        if (i >= coins[j]) {
          const val = dp[i - coins[j]];
          if (val + 1 < dp[i]) dp[i] = val + 1;
        }
      }
    }

    count += dp[M];
    process.stdout.write(count + "\n");
  }
}

solve();

File: BeeCrowd/1034-Ice-Statues-Festival/solution.cpp
/**
 * BeeCrowd 1034 - Ice Statues Festival
 * Author: [Your Name/Handle]
 * Algorithm: Hybrid Greedy + Unbounded Knapsack
 * Complexity: O(C * N) where C is constant (~10,000)
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

const int INF = 1e9;

void solve()
{
  int T;
  // Fast I/O is crucial for competitive C++
  if (!(cin >> T))
    return;

  // We reuse this buffer to avoid memory allocation overhead per test case.
  // Max needed size is approx 100 * 100 = 10,000.
  // 20,000 is a safe upper bound.
  vector<int> dp(20000);

  while (T--)
  {
    int N, M;
    cin >> N >> M;

    vector<int> coins(N);
    for (int i = 0; i < N; ++i)
    {
      cin >> coins[i];
    }

    // 1. Sort Descending
    // We need the largest coins to calculate the "Safe Limit"
    sort(coins.begin(), coins.end(), greater<int>());

    int max_coin = coins[0];
    int count = 0;

    // 2. Dynamic Limit Heuristic
    // The pattern of optimal coins stabilizes after Largest * 2nd_Largest.
    // We only need to DP the "unstable" remainder.
    int safe_limit = 0;
    if (N >= 2)
    {
      safe_limit = max_coin * coins[1];
    }

    // 3. Greedy Reduction
    // If M is huge, peel off layers of the largest coin
    // until M is close to our safe_limit.
    if (M > safe_limit)
    {
      int reduce_amount = M - safe_limit;
      int num_max_coins = reduce_amount / max_coin;
      if (num_max_coins > 0)
      {
        count += num_max_coins;
        M -= num_max_coins * max_coin;
      }
    }

    // 4. Dynamic Programming
    // Solve the small remainder using standard Unbounded Knapsack

    // Reset DP table (only up to the new M)
    fill(dp.begin(), dp.begin() + M + 1, INF);
    dp[0] = 0;

    for (int i = 1; i <= M; ++i)
    {
      for (int j = 0; j < N; ++j)
      {
        if (i >= coins[j])
        {
          if (dp[i - coins[j]] != INF)
          {
            dp[i] = min(dp[i], dp[i - coins[j]] + 1);
          }
        }
      }
    }

    count += dp[M];
    cout << count << "\n";
  }
}

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  solve();
  return 0;
}
